src/Magnum/DartIntegration/ConvertShapeNode.h

```diff
@@ -0,0 +1,82 @@
+#define Magnum_DARTIntegration_ConvertShapeNode_h
+/*
+    This file is part of Magnum.
+
+    Copyright © 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018
+              Vladimír Vondruš <mosra@centrum.cz>
+    Copyright © 2018 Konstantinos Chatzilygeroudis <costashatz@gmail.com>
+
+    Permission is hereby granted, free of charge, to any person obtaining a
+    copy of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included
+    in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+*/
+
+/** @file
+ * @brief Function @ref Magnum::DartIntegration::convertShapeNode()
+ * @brief Class @ref Magnum::DartIntegration::VisualData
+ */
+
+#include <vector>
+
+#include <Corrade/Containers/Optional.h>
+
+#include <Magnum/DartIntegration/DartObject.h>
+#include <Magnum/DartIntegration/visibility.h>
+
+namespace Magnum {namespace Trade{
+    class MeshData3D;
+    class PhongMaterialData;
+    template<UnsignedInt dimensions> class ImageData;
+    typedef ImageData<2> ImageData2D;
+    class TextureData;
+}}
+
+namespace Magnum { namespace DartIntegration {
+/**
+@brief VisualData struct
+*/
+struct VisualData {
+    /** mesh information */
+    Trade::MeshData3D* _meshData;
+    /** material information */
+    Trade::PhongMaterialData* _matData;
+    /** ImageData2D (vector) */
+    std::vector<Trade::ImageData<2>*> _imgData;
+    /** TextureData (vector) */
+    std::vector<Trade::TextureData*> _texData;
+};
+
+/**
+@brief Convert Dart ShapeNode to @ref VisualData that can be processed by Magnum
+@param shapeNode   Dart ShapeNode to convert
+
+Returns `Corrade::Containers::NullOpt` if the shape of the given ShapeNode is not supported.
+*/
+Corrade::Containers::Optional<VisualData> MAGNUM_DARTINTEGRATION_EXPORT convertShapeNode(dart::dynamics::ShapeNode* shapeNode);
+
+/**
+@brief Convert DartObject to @ref VisualData that can be processed by Magnum
+
+See @ref convertShape(dart::dynamics::ShapeNode* shapeNode)
+for more information.
+*/
+Corrade::Containers::Optional<VisualData> MAGNUM_DARTINTEGRATION_EXPORT convertShapeNode(DartObject* object);
```

Comment:
> Could these two functions take a reference instead of a pointer? So the nullptr check responsibility is moved to the user of the API.

I think they can yes. But why is it better to move the nullptr check to the user?

Reply:
Because the users usually know better if given pointer is valid or not, so they can in many cases omit the check altogether. 

I see it as either needing to check for `nullptr` *before* or handling the return value being `NullOpt` *after* (and then maybe wondering what was the cause). In this case I think the former is simpler and thus better.

Reply:
> I see it as either needing to check for nullptr before or handling the return value being NullOpt after (and then maybe wondering what was the cause). In this case I think the former is simpler and thus better.

Sounds good! It's also better for everything Magnum related to have similar styles.. :smile: